Redis Function

MeirShpilraien opened the pull request:
# Redis Function
This PR added the Redis Functions capabilities that were suggested on https://github.com/redis/redis/issues/8693. The PR also introduce a big refactoring to the current Lua implementation (i.e `scripting.c`). The main purpose of the refactoring is to have better code sharing between the Lua implementation that exists today on Redis (`scripting.c`) and the new Lua engine that is introduced on this PR. The refactoring includes code movements and file name changes as well as some logic changes that need to be carefully reviewed. To make the review easier, the PR was split into multiple commits. Each commit is deeply described later on but the main concept is that some commits are just moving code around without making any logical changes, those commits are less likely to cause any issues or regressions and can be reviewed fast. Other commits, which perform code and logic changes, need to be reviewed carefully, but those commits were created after the code movements so it's pretty easy to see what was changed. To sum up, it is highly recommended to review this PR commit by commit as it will be easier to see the changes, it is also recommended to read each commit description (written below) to understand what was changed on the commit and whether or not it's just a huge code movement or a logic changes.

## Terminology
Currently, the terminology in Redis is not clearly defined. Scripts refer to Lua scripts and eval also refers only to Lua. Introducing Redis Function requires redefining those terms to be able to clearly understand what is been discussed on each context.
* eval - legacy Lua script implementation.
* Function - new scripting implementation (currently implemented in Lua but in the future, it might be other languages like javascript).
* Engine - the component that is responsible for executing functions.
* Script - Function or legacy Lua (executed with `eval` or `evalsha`)

## Refactoring New Structure
Today, the entire scripting logic is located on `scripting.c`. This logic can be split into 3 main groups:
1. Script management - responsible for storing the scripts that were sent to Redis and retrieving them when they need to be run (base on the script sha on the current implementation).
2. Script invocation - invoke the script given on `eval` or `evalsha` command (this part includes finding the relevant script, preparing the arguments, ..)
3. Interact back with Redis (command invocation)

Those 3 groups are tightly coupled on `scripting.c`. Redis Functions also need to use those groups logics, for example,  to interact back with Redis or to execute Lua code. The refactoring attempts to split those 3 groups and define APIs so that we can reuse the code both on legacy Lua scripts and Redis Functions.

In order to do so we define the following units:
1. script.c: responsible for interaction with Redis from within a script.
2. script_lua.c: responsible to execute Lua code, uses `script.c` to interact with Redis from within the Lua code.
3. function_lua.c: contains the Lua engine implementation, uses `script_lua.c` to execute the Lua code.
4. functions.c: Contains Redis Functions implementation (`FUNCTION` command,), uses `functions_lua.c` if the function it wants to invoke needs the Lua engine.
4. eval.c: the original `scripting.c` contains the Lua legacy implementation and was refactored to use `script_lua.c` to invoke the Lua code.

## Commits breakdown
Notice: Some small commits are omitted from this list as they are small and insignificant (for example build fixes)
### First commit - code movements
This commit rename `scripting.c` -> `eval.c` and introduce the new `script_lua.c` unit. The commit moves relevant code from `eval.c` (`scripting.c`) to `script_lua.c`, the purpose of moving the code is so that later we will be able to re-use the code on the Lua engine (`function_lua.c`). The commit only moves the code without modifying even a single line, so there is a very low risk of breaking anything and it also makes it much easier to see the changes on the following commits.
Because the commit does not change the code (only moves it), it does not compile. But we do not care about it as the only purpose here is to make the review processes simpler.

### Second commit - move legacy Lua variables into `eval.c`
Today, all Lua-related variables are located on the server struct. The commit attempt to identify those variable and take them out from the server struct, leaving only script related variables (variables that later need to be used also by engines)
The following variable where renamed and left on the server struct:
   * lua_caller 			-> script_caller
   * lua_time_limit 		-> script_time_limit
   * lua_timedout 		-> script_timedout
   * lua_oom 			-> script_oom
   * lua_disable_deny_script 	-> script_disable_deny_script
   * in_eval			-> in_script

The following variables where moved to lctx under eval.c
   * lua
   * lua_client
   * lua_cur_script
   * lua_scripts
   * lua_scripts_mem
   * lua_replicate_commands
   * lua_write_dirty
   * lua_random_dirty
   * lua_multi_emitted
   * lua_repl
   * lua_kill
   * lua_time_start
   * lua_time_snapshot
   
This commit is in a low risk of introducing any issues and it is just moving variables around and not changing any logic.

### Third commit - introducing script unit
This commit introduces the `script.c` unit. Its purpose (as described above) is to provide an API for scripts to interact with Redis. Interaction includes mostly executing commands, but also other functionalities. The interaction is done using a `ScriptRunCtx` object that needs to be created by the user and initialized using `scriptPrepareForRun`. A detailed list of functionalities expose by the unit:
1. Calling commands (including all the validation checks such as
   acl, cluster, read only run, ...)
2. Set Resp
3. Set Replication method (AOF/REPLICATION/NONE)
4. Call Redis back on long-running scripts to allow Redis to reply to clients and perform script kill

The commit introduces the new unit and uses it on eval commands to interact with Redis.

### Fourth commit - Moved functionality of invoke Lua code to `script_lua.c`
This commit moves the logic of invoking the Lua code into `script_lua.c` so later it can be used also by Lua engine (`function_lua.c`). The code is located on `callFunction` function and assumes the Lua function already located on the top of the Lua stack. This commit also change `eval.c` to use the new functionality to invoke Lua code.

### Fith commit - Added Redis Functions unit (`functions.c`) and Lua engine (`function_lua.c`)
Added Redis Functions unit under `functions.c`, included:
1. FUNCTION command:
     * FUNCTION CREATE
     * FUNCTION CALL
     * FUNCTION DELETE
     * FUNCTION KILL
     * FUNCTION INFO
     * FUNCTION STATS
2. Register engines

In addition, this commit introduces the first engine that uses the Redis Functions capabilities, the Lua engine (`function_lua.c`)

## API Changes
### `lua-time-limit` 
configuration was renamed to `script-time-limit` (keep `lua-time-limit` as alias for backward compatibility).

### Error log changes
When integrating with Redis from within a Lua script, the `Lua` term was removed from all the error messages and instead we write only `script`. For example:
`Wrong number of args calling Redis command From Lua script` -> `Wrong number of args calling Redis command From script`

### `info memory` changes:
Before stating all the changes made to memory stats we will try to explain the reason behind them and what we want to see on those metrics:
* memory metrics should show both totals (for all scripting frameworks), as well as a breakdown per framework / vm.
* The totals metrics should have "human" metrics while the breakdown shouldn't.
* We did try to maintain backward compatibility in some way, that said we did make some repurpose to existing metrics where it looks reasonable.
* We separate between memory used by the script framework (part of redis's used_memory), and memory used by the VM (not part of redis's used_memory)

A full breakdown of `info memory` changes:
* `used_memory_lua` and `used_memory_lua_human` was deprecated, `used_memory_vm_eval` has the same meaning as `used_memory_lua`
* `used_memory_scripts` was renamed to `used_memory_scripts_eval`
* `used_memory_scripts` and `used_memory_scripts_human` were repurposed and now return the total memory used by functions and eval (not including vm memory, only code cache, and structs).
* `used_memory_vm_function` was added and represents the total memory used by functions vm's
* `used_memory_functions` was added and represents the total memory by functions (not including vm memory, only code cache, and structs)
* `used_memory_vm_total` and `used_memory_vm_total_human` was added and represents the total memory used by vm's (functions and eval combined)

### `functions.caches`
`functions.caches` field was added to `memory stats`, representing the memory used by engines that are not functions (this memory includes data structures like dictionaries, arrays, ...)

## New API
### FUNCTION CREATE

Usage: FUNCTION CREATE `ENGINE` `NAME` `[REPLACE]` `[DESC <DESCRIPTION>]`  `<CODE>`

* `ENGINE` - The name of the engine to use to create the script.
* `NAME` - the name of the function that can be used later to call the function using `FUNCTION CALL` command.
* `REPLACE` - if given, replace the given function with the existing function (if exists).
* `DESCRIPTION` - optional argument describing the function and what it does
* `CODE` - function code.

The command will return `OK` if created successfully or error in the following cases:
* The given engine name does not exist
* The function name is already taken and `REPLACE` was not used.
* The given function failed on the compilation.

### FCALL and FCALL_RO

Usage: FCALL/FCALL_RO `NAME` `NUM_KEYS key1 key2` … ` arg1 arg2`

Call and execute the function specified by `NAME`. The function will receive all arguments given after `NUM_KEYS`. The return value from the function will be returned to the user as a result.

* `NAME` - Name of the function to run.
* The rest is as today with EVALSHA command.

The command will return an error in the following cases:
* `NAME` does not exist
* The function itself returned an error.

The `FCALL_RO` is equivalent to `EVAL_RO` and allows only read-only commands to be invoked from the script.

### FUNCTION DELETE

Usage: FUNCTION DELETE `NAME`

Delete a function identified by `NAME`. Return `OK` on success or error on one of the following:
* The given function does not exist

### FUNCTION INFO

Usage: FUNCTION INFO `NAME` [WITHCODE]

Return information about a function by function name:
* Function name
* Engine name
* Description 
* Raw code (only if WITHCODE argument is given)

### FUNCTION LIST

Usage: FUNCTION LIST

Return general information about all the functions:
* Function name
* Engine name
* Description

### FUNCTION STATS

Usage: FUNCTION STATS

Return information about the current running function:
* Function name
* Command that was used to invoke the function
* Duration in MS that the function is already running

If no function is currently running, this section is just a RESP nil.

Additionally, return a list of all the available engines.

### FUNCTION KILL

Usage: `FUNCTION KILL`

Kill the currently executing function. The command will fail if the function already initiated a write command.

## TODO
- [x] Handle short read (add tests)
- [x] Improve testing

## Community requests to handle on following PR's
* Allow loading functions from an external file on startup
* Define and implement Function arguments parsing (named arguments/positional arguments)
* Define and implement code reuse between functions
* Add JS engine

## Community requests handled on this PR's
* FUNCTION RUNNING command, to check to current running function: https://github.com/redis/redis/issues/8693#issuecomment-859957474
Handled with FUCTION STATS command

* Get the source code of a function: https://github.com/redis/redis/issues/8693#issuecomment-859957474
Handled with FUCTION INFO command

## Community requests to consider for future planning
* Memory limit on functions engines.
* Trigger functions on events: https://github.com/redis/redis/issues/8693#issuecomment-812481386
* Blocking operations inside functions: https://github.com/redis/redis/issues/8693#issuecomment-815007678
* Background function execution: https://github.com/redis/redis/issues/8693#issuecomment-859957474
* Rollback on timeouts: https://github.com/redis/redis/issues/8693#issuecomment-816630862

Note: Function creation/deletion is replicated to AOF  but AOFRW is not implemented sense its going to be removed: https://github.com/redis/redis/issues/9794

oranagra commented:
@redis/core-team please review / approve and state what's missing or needs a change.
there are a few additional minor changes that should be done on this PR before being merged, but i think the majority of the additional changes can come in a few followup PRs.

eduardobr commented:
@MeirShpilraien I noticed the functionsCtxGetCurrent gets a global value. Does it make sense that it could live within the redisDb struct? Same style as slots_to_keys, that was moved there recently.
Reason for that is that passing around, initializing, swapping and destroying server.db and tempDb + related objects is cleaner and more explicit when the related objects come together (at least things that share same life cycle).
Methods like functionsCtxSwapWithCurrent would be called or have the logic in the main function that swaps whole db.

oranagra commented:
@eduardobr the problem is that functions are global (not per db, in a multi-db server).
unlike the cluster slots mapping, which is global, but only valid in a single db configuration.
that's why we had to create `rdbLoadingCtx`, if we had that one before, maybe we would have put the slots mapping there too.

Note, i did review this code before being posted publicly, and did attempt to follow the cluster slots footsteps as much as i could.
if you still think there's a possibility for an additional cleanup, please suggest it.

sundb commented on src/eval.c:
```diff
@@ -0,0 +1,1637 @@
+sds ldbCatStackValue(sds s, lua_State *lua, int idx);
+
+/* Lua context */
+struct luaCtx{
```
```suggestion
struct luaCtx {
```

sundb commented on src/eval.c:
```diff
@@ -0,0 +1,1637 @@
+struct luaCtx{
+    lua_State *lua; /* The Lua interpreter. We use just one for all clients */
+    client *lua_client;   /* The "fake client" to query Redis from Lua */
+    char* lua_cur_script; /* SHA1 of the script currently running, or NULL */
```
```suggestion
    char *lua_cur_script; /* SHA1 of the script currently running, or NULL */
```

yossigo reviewed (approved):
LGTM, but did not do a full code review of the latest version.

itamarhaber reviewed (approved):
Approving the feature and API, did not do a code review.

madolson commented on tests/unit/functions.tcl:
```diff
@@ -0,0 +1,248 @@
+start_server {tags {"scripting"}} {
+	test {FUNCTION - Basic usage} {
```
You have a tab here.

madolson commented on src/script.c:
```diff
@@ -0,0 +1,422 @@
+    return C_OK;
+}
+
+static int scriptVerifyOOM(scriptRunCtx *r_ctx, char **err) {
+    /* If we reached the memory limit configured via maxmemory, commands that
+     * could enlarge the memory usage are not allowed, but only if this is the
+     * first write in the context of this script, otherwise we can't stop
+     * in the middle. */
```
We have a known issue where a script can execute a delete first, and then continue to use memory and grow. It seems like we should fix that initially here.

MeirShpilraien replied:
Yes we actually have an idea to handle this issue and will share on following PR's/issues. For this PR I wanted to keep the behaviour untouched.

madolson replied:
Sounds good, we should just make sure it's fixed before 7, since it seems like we'll need some type of breaking change.

madolson commented on src/script.c:
```diff
@@ -0,0 +1,422 @@
+
+    if (cmd->flags & CMD_WRITE) {
+        /* signify that we already change the data in this execution */
+        r_ctx->flags |= SCRIPT_WRITE_DIRTY;
```
It looks like we do this today, but just because a command is a write, doesn't mean it dirties the dataset. I suppose this means we might send a multi-exec filled with commands that don't execute. Not a correctness problem I presume, just weird.

MeirShpilraien replied:
I am not sure what you mean here, do you refer to this issue: https://github.com/redis/redis/issues/8478?

madolson replied:
No, if we send a write command we mark the script as dirty. We don't need to though, we only need to care if the write command actually does something to the dataset. I'll open an separate issue for this, it's not related but something I noticed.

MeirShpilraien replied:
I see, so I guess I can resolve this one?

oranagra replied:
yes, this is not related to here, let's add it to #8478

madolson commented on src/script.c:
```diff
@@ -0,0 +1,422 @@
+ * up to the engine to take and parse.
+ * The err out variable is set only if error occurs and describe the error.
+ * If err is set on reply is written to the r_ctx client. */
+void scriptCall(scriptRunCtx *r_ctx, robj* *argv, int argc, sds *err) {
```
Can we call r_ctx, "run_ctx", while reading this I keep thinking it must be a "redis context".

madolson commented on src/rdb.c:
```diff
@@ -1239,6 +1240,25 @@ int rdbSaveRio(rio *rdb, int *error, int rdbflags, rdbSaveInfo *rsi) {
     if (rdbSaveModulesAux(rdb, REDISMODULE_AUX_BEFORE_RDB) == -1) goto werr;
 
+    /* save functions */
+    dict* functions = functionsGet();
+    dictIterator *iter = dictGetIterator(functions);
```
You need to make up your mind about where the pointer *s go :)

MeirShpilraien replied:
At first I did everything like this one (pointer close to the type). But then I saw that the conversion in Redis is that the pointer is next to the variable name and I changed. Must a missed some. Will fix and re-scan the code, thanks.

madolson commented on src/server.c:
```diff
@@ -5185,14 +5249,14 @@ void populateCommandMovableKeys(struct redisCommand *cmd) {
  * if C_ERR is returned the client was destroyed (i.e. after QUIT). */
 int processCommand(client *c) {
-    if (!server.lua_timedout) {
+    if (!server.script_timedout) {
```
```suggestion
    if (!isScriptTimedOut()) {
```

madolson commented on src/server.c:
```diff
@@ -3287,6 +3330,24 @@ void blockingOperationEnds() {
     }
 }
 
+int isScriptTimedout() {
+    return server.script_timedout;
+}
```
Should this be moved into script.(c|h)? Perhaps a better time for a little more modularity. We can pull the stuff out of server.h as well.

MeirShpilraien replied:
Good point, will move it.

MeirShpilraien replied:
Actually, re-looking at this, it is looking at the server struct so maybe we should keep it here?

madolson replied:
Maybe it shouldn't be in server struct at all then?

MeirShpilraien replied:
I agree, added another PR and moved the entire timeout handling into `script.c`

madolson commented on src/functions.c:
```diff
@@ -0,0 +1,445 @@
+ * ENGINE NAME     - name of the engine to use the run the function
+ * FUNCTION NAME   - name to use to invoke the function
+ * REPLACE         - optional, replace existing function
+ * DESC            - optional, function description
```
I generally dislike ambiguous acronyms (like DESC) can we just put DESCRIPTION ?

madolson commented on src/server.c:
```diff
@@ -3530,8 +3591,10 @@ void createSharedObjects(void) {
+    shared.slowevalerr = createObject(OBJ_STRING,sdsnew(
         "-BUSY Redis is busy running a script. You can only call SCRIPT KILL or SHUTDOWN NOSAVE.\r\n"));
+    shared.slowscripterr = createObject(OBJ_STRING,sdsnew(
+        "-BUSY Redis is busy running a script. You can only call FUNCTION KILL or SHUTDOWN NOSAVE.\r\n"));
```
We could also tell the user what script is being run because these have names, that seems useful?

MeirShpilraien replied:
Currently the run_ctx is shared between eval and functions and it is not aware about the functions which is currently running (his purpose is to interact with Redis from within a script). I can add optional metadata that can be added there and we can put the running function in this metadata. Then we will be able to reply with the function name, WDYT?

oranagra replied:
we only recently added something like that for EVAL, we print the name to the log.
either way, i don't think it should be part of the error message.
maybe a different introspection command that is allowed to run during a busy script?

madolson replied:
I'm fine either way, I just think the user should have a way to know what "Function" is running.

MeirShpilraien replied:
We discussed it on the issue (https://github.com/redis/redis/issues/8693#issuecomment-860160430), @oranagra mentioned that its not enough to know the function that is running because it might be change by the time we will send the `FUNCTION KILL`. Even if we will allow to kill only if the function name is equal to a given parameter, it still might have finished and restarted by the time we will send the command.
Regardless, we can add `FUNCTION RUNNING` that will return the current running function (as mentioned on the PR message), I am just not sure it will help the user kill the "right" function. Let me know what you think.

oranagra replied:
I agree, it should be some sort of introspection command, (not part of the error)
maybe instead of FUNCTION RUNNING, we can add some FUNCTION STATUS or STATS that will print some general info on all engines (unlike FUNCTION INFO which is on one function, right?).
i.e. something similar to MEMORY STATS and INFO.
there was can put all soft of general details, including the currently running function, it's args, and execution duration.

MeirShpilraien replied:
Added FUNCTION STATS command, update the top comment with the relevant info.

madolson commented on src/server.c:
```diff
@@ -3530,8 +3591,10 @@ void createSharedObjects(void) {
         "-LOADING Redis is loading the dataset in memory\r\n"));
-    shared.slowscripterr = createObject(OBJ_STRING,sdsnew(
+    shared.slowevalerr = createObject(OBJ_STRING,sdsnew(
         "-BUSY Redis is busy running a script. You can only call SCRIPT KILL or SHUTDOWN NOSAVE.\r\n"));
```
Based on terminology, it's non-obvious why SCRIPT KILL wouldn't also apply to functions. You said a script is either an eval or function.

MeirShpilraien replied:
Also had this discussion with @oranagra , the reason is that it might confuse the user because `SCRIPT LOAD` for example is not apply to function.

oranagra replied:
yes, i also gave that comment, and after some discussions retracted it.
the old terminology was mixing [lua, eval, script], in some cases it suites us, in orders we had to rename or break.

in this one, we can choose to support SCRIPT KILL, and then drop the FUNCTION KILL command.
and then we can also use just one error message rather than two, or accept the current code which regards the SCRIPT command as an EVAL manager.

maybe another new idea to solve this is merge the entire FUNCTION command into the SCRIPT command.
so instead of FUNCTION CREATE, we'll have SCRIPT FCREATE (similar to the FCALL).
then when the entire SCRIPT command is about both functions and eval, the terminology problem is solved.
i.e. obviously some of the sub-commands will be only applicable to some cases, but it might not be that bad.

p.s. i'm not sure i like the idea i just suggested. maybe the current code is better.

oranagra replied:
@yossigo maybe you can settle this?

yossigo replied:
@oranagra I also don't like the last idea.

Generally I don't think scripts and functions should be confused, but `-BUSY` does imply a more general state, so it's a bit weird to require the user to determine what keeps Redis busy and then decide if `SCRIPT KILL` or `FUNCTION KILL` should be executed.

Maybe support both and let users use them interchangeably?

oranagra replied:
@yossigo so what you mean is that you don't like to introduce the function management commands as sub-commands of SCRIPT? and you're ok with letting SCRIPT remain dedicated for the EVAL scripts (despite the fancy terminology we creates which states that a "script" can refer to both functions and eval)?

and you're suggesting to just tackle the KILL problem as a special case, in which we don't wanna force the user to be able to recognize them apart (it's the same error code), and instead we'll just have two KILL sub-commands, interchangeable (either one can kill both functions and eval)?

did i get your thought correctly?

MeirShpilraien replied:
After discussion with @oranagra and @yossigo the decision is to keep the API separate.

oranagra replied:
some more details about our discussion.

First, about considering moving the FUNCTION command logic to SCRIPT (since our new terminology dictates that both "EVAL" and "Functions" are "scripts), we concluded that in this case it'll cause too much confusion (e.g. SCRIPT LOAD will obviously remain as is), and it's ok to leave the administrative parts of EVAL in SCRIPT, and the ones of FCALL in FUNCTION.

Secondly, regarding KILL, and BUSY:
We concluded that the BUSY response is more about telling the client that the server is busy and the command should be retried later, than about a way to KILL the blocker.
I even remember a discussion with Salvatore about EXEC returning BUSY without resetting the MULTI state so that it can be retried later (we abandoned that since it's dangerous, and it always fails with EXECABORT).
Moreover, we have a plan to soon let modules play that BUSY game too (see #9655), and they're obviously not gonna get killed by SCRIPT KILL.

p.s SCRIPT KILL is in many time a dead end anyway (responds with UNKILLABLE)

madolson commented on src/functions.c:
```diff
@@ -0,0 +1,445 @@
+"    Create a new function with the given function name and code.",
+"DELETE <FUNCTION NAME>",
+"    Delete the given function.",
+"INFO",
```
KILL?

madolson commented on src/functions.c:
```diff
@@ -0,0 +1,445 @@
+"    For each function, print the following information about the function:",
+"    * Function name",
+"    * The engine used to run the function",
+"    * Function description",
```
Why aren't we also printing back the raw function code? Do we think it's too much memory to store?

MeirShpilraien replied:
We actually do store the raw function code to be able to save it to RDB. I thought that returning it on info might be to much and also not sure we want to commit to always be able to return the raw function code (it also might be meaningless on other engines in the future, like web-assembly for example). WDYT?

oranagra replied:
maybe we want to add another sub-command for that, better leave it outside of INFO, to make it short.
i.e we can add FUNCTION GET and FUNCTION ALL and alike.

i don't think it'll be a problem for WASM, we'll just spit back the code we got from the user.
either way we'll need to persist them, so even if we only store it in the engine, we have to have a way to get it back.

MeirShpilraien replied:
@oranagra I think it will be irrelevant with WASM because the "code" we will get is a binary blob, nothing readable or meaningful.

oranagra replied:
first, everything can be desassembled 8-)
but also, people can get it in order to load it into another redis (or another system).
i think it is ok if we spit out what we got as input.

MeirShpilraien replied:
Its actually on the following PR list:
> Get the source code of a function

So I will come with a proposal and implementation.

madolson replied:
I think having a separate subcommand with more info sounds like the right approach.

MeirShpilraien replied:
Should we do it on this PR or open an issue with API definition and then handle it on following PR?

oranagra replied:
IMHO if you have time to kill, and it doesn't add another major approval barrier to this PR, you can go ahead and add it.

MeirShpilraien replied:
Added the option to specify function name on FUNCTION INFO command that also return the raw code. Updated the top comment accordingly.

oranagra replied:
does that mean that `FUNCTION INFO`, and `FUNCTION INFO <name>` return completely different result?
if it does, i don't like that.

MeirShpilraien replied:
So you suggest a new sub-command? FUNCTION CODE?

oranagra replied:
i think we should have
* `FUNCTION STATS` - general info on the engine and some counters (no per function details)
* `FUNCTION LIST` - listing all function and some metadata on each (excluding the function code)
* `FUNCTION INFO <name> [WITHCODE]` - detailed info on the function, optionally including it's code.

MeirShpilraien replied:
Updated the code accordingly, also updated top comment.

madolson commented on src/functions.c:
```diff
@@ -0,0 +1,445 @@
+#include "dict.h"
+#include "adlist.h"
+#include "atomicvar.h"
+#include "cluster.h"
```
Why is cluster needed here? I didn't see anything cluster specific in this file, and conceptually it doesn't make sense to me.

MeirShpilraien replied:
Mistake, thanks.

madolson reviewed (commented):
(Sorry for random comments scattered, I was just making notes on stuff I saw. I mostly was trying to understand the interface. It looked good, so I trust the functionality is implemented as expected)

I want to better understand why we think the right approach here is storing functions in the RDB as opposed to storing them in the conf file or a separate file. I view functions more analogous to ACLs than keyspace data, that they are properties of the cluster and not necessarily specific to the RDB data. This is basically completely broken for workloads without persistence, like caching, which seems relevant. I would expect having a well defined mechanism outside of the RDB to exist.

Also are we fixing this? https://github.com/redis/redis/issues/8478 for functions?

madolson commented on tests/unit/functions.tcl:
```diff
@@ -0,0 +1,248 @@
+        }
+
+        test {FUNCTION - creation is replicated to replica} {
+        	r function create LUA test DESC {some description} {return 'hello'}
+        	wait_for_condition 50 100 {
+        		[r -1 function info] eq {{name test engine LUA desc {some description}}}
+        	} else {
+                fail "Failed waiting for function to replicate to replica"
```
More tabs.

MeirShpilraien replied:
replace all tabs with spaces

MeirShpilraien commented:
@madolson thanks for the review. Fixed most of the comments (pushed a new commit) and I added comments where I needed more input about how to proceed. 

Regarding pre-load of functions (not from persistence). This is listed as first task to handle on following PR's. I am working on the design and will share in a few days. Once you (the core team) will approve the design I will work on the implementation.

Regarding this https://github.com/redis/redis/issues/8478, I have another design coming to allow better code sharing and I hope to address it (or at least suggest a possible solution) there. Will share soon.

madolson commented:
@MeirShpilraien My point is that I'm not convinced RDB is the right place to be storing functions, or at least we should take a deeper look into the RDB storage to think about how it interacts with other features. One example is cluster mode, when you want to split a cluster, you add N new nodes with the same configs of the other nodes and then migrate the keyspace data off. This won't include functions until we add them to configs, but then you run into the issue where they need to only be loaded on primaries and replicated (or perhaps we allow primaries and replicas to diverge?).

That's why this still feels more analogous to the config file or to ACLs than it does to keyspace data. The only real benefit here is that storing it in the RDB, is actually really convenient. But it would also be convenient for ACLs, so maybe the right answer is we should be storing them both in the RDB.

MeirShpilraien commented:
@madolson the entire point of functions is that they are replicated and persisted so you will have them on fail-overs for example (this is also how we defined it in the issue https://github.com/redis/redis/issues/8693). I do agree though that we need a way to allow users to also load them from config in case the user runs without persistence for example. Hope to publish my design for that today/tomorrow and we will be able to discuss it.

> (Also, I didn't actually review the code but I assume we replicate by the effects of the script and not the script itself. If that isn't the case, then having replica and primary agree on the state of a function matter more)

Yes, on functions we replicate only by effect.

madolson commented:
@MeirShpilraien Agree about having them available on failover, if they were stored in a config file they would also be available though.

sundb commented on src/functions.c:
```diff
@@ -0,0 +1,446 @@
+    functionInfo *fi = dictFetchValue(functions->functions, function_name);
+    if (fi) {
+        if (replace) {
+            engineFunctionFree(fi, functions);
```
Should free `fi` after `engine->create` successfully?

MeirShpilraien replied:
No, we should only free it if it already exists and we replacing it.

sundb replied:
I mean that when we create a function that fails, we shouldn't delete the old one.

```sh
$ function create lua test replace 'return "hello"'
$ function info
1) 1) "name"
   2) "test"
   3) "engine"
   4) "LUA"
   5) "description"
   6) (nil)
$ function create lua test replace 'error'
(error) ERR Error compiling function: user_function:1: '=' expected near '<eof>'
$ function info
(empty array)
```

Expectation

```sh
$ function create lua test replace 'return "hello"'
$ function info
1) 1) "name"
   2) "test"
   3) "engine"
   4) "LUA"
   5) "description"
   6) (nil)
$ function create lua test replace 'error'
(error) ERR Error compiling function: user_function:1: '=' expected near '<eof>'
$ function info
1) 1) "name"
   2) "test"
   3) "engine"
   4) "LUA"
   5) "description"
   6) (nil)
```

MeirShpilraien replied:
Ohh, nice catch, thank you.

MeirShpilraien replied:
Added a commit that fixes it.

MeirShpilraien replied:
@sundb please let me know if you agree with the fix and if I can resolve this.

sundb replied:
@MeirShpilraien It's OK.

sundb commented on src/functions.h:
```diff
@@ -0,0 +1,125 @@
+} functionInfo;
+
+int functionsRegisterEngine(const char *engine_name, engine *engine_ctx);
+int functionsCreateWithFunctionCtx(sds function_name,sds engine_name, sds desc, sds code,
```
```suggestion
int functionsCreateWithFunctionCtx(sds function_name, sds engine_name, sds desc, sds code,
```

MeirShpilraien replied:
Fixed

sundb commented on src/functions.c:
```diff
@@ -0,0 +1,447 @@
+
+    functionInfo *fi = dictFetchValue(functions->functions, function_name);
+    if (fi && !replace) {
+		*err = sdsnew("Function already exists");
+		return C_ERR;
```
Incorrect tab indent.

MeirShpilraien replied:
Yeah, just saw it, thanks.

sundb commented on src/functions.c:
```diff
@@ -0,0 +1,447 @@
+
+    if (fi) {
+    	/* free the already existing function as we are going to replace it */
+		engineFunctionFree(fi, functions);
+	}
```
Another tab indent.

MeirShpilraien replied:
:+1:

sundb commented on src/server.c:
```diff
@@ -471,6 +472,25 @@ struct redisCommand scriptSubcommands[] = {
+    {"create",functionsCreateCommand,-5,
+     "may-replicate no-script @scripting"},
+
+    {"delete",functionsDeleteCommand,-3,
```
```suggestion
    {"delete",functionsDeleteCommand,3,
```

sundb commented on src/functions.c:
```diff
@@ -0,0 +1,447 @@
+ * 3. update cache memory
+ */
+static void engineFunctionRegister(functionInfo *fi, functionsCtx *functions) {
+    dictAdd(functions->functions, fi->name, fi);
```
Although `functionsCreateWithFunctionCtx` has already deleted the key, I think it's safer to add an assertion.
```suggestion
    assert(dictAdd(functions->functions, fi->name, fi) == DICT_OK);
```

MeirShpilraien replied:
I do not like putting code that should run inside an assert (in case we want to run without asserts someday), will save the return value and assert it.

sundb replied:
Yes, it saves a line of code when writing in assert if we don't need the return value.

MeirShpilraien replied:
I mean that I want to save the return value in a variable and then assert, this way if someone compiles without asserts it will still work:
```
int res = dictAdd(functions->functions, fi->name, fi);
assert(res == DICT_OK);
```

sundb commented on src/functions.c:
```diff
@@ -0,0 +1,447 @@
+
+/* Create a new functions ctx */
+functionsCtx* functionsCtxCreate() {
+    functionsCtx* ret = zmalloc(sizeof(functionsCtx));
```
```suggestion
    functionsCtx *ret = zmalloc(sizeof(functionsCtx));
```

sundb commented on src/functions.c:
```diff
@@ -0,0 +1,447 @@
+        NULL,              /* val dup */
+        dictSdsKeyCompare, /* key compare */
+        dictSdsDestructor, /* key destructor */
+        NULL,              /* val destructor */
```
It seems that it could go here, which would eliminate the need to call it when deleted.
```suggestion
        engineFunctionDispose,              /* val destructor */
```

sundb commented on src/functions.h:
```diff
@@ -0,0 +1,126 @@
+ */
+
+#ifndef SRC_FUNCTIONS_H_
+#define SRC_FUNCTIONS_H_
```
Would it be better to keep the original style?
```suggestion
#define __FUNCTIONS_H_
```

sundb commented on src/rdb.c:
```diff
@@ -1239,6 +1240,25 @@ int rdbSaveRio(rio *rdb, int *error, int rdbflags, rdbSaveInfo *rsi) {
+        if (rdbSaveRawString(rdb, (unsigned char *) fi->ei->name, sdslen(fi->ei->name)) == -1) goto werr;
+        if (fi->desc) {
+            if (rdbSaveLen(rdb, 1) == -1) goto werr; /* desc exists */
+            if (rdbSaveRawString(rdb, (unsigned char *) fi->desc, sdslen(fi->name)) == -1) goto werr;
```
```suggestion
            if (rdbSaveRawString(rdb, (unsigned char *) fi->desc, sdslen(fi->desc)) == -1) goto werr;
```

sundb commented on src/script.c:
```diff
@@ -0,0 +1,465 @@
+static scriptRunCtx *curr_run_ctx = NULL;
+
+static void exitScriptTimedoutMode(scriptRunCtx *run_ctx) {
+	serverAssert(run_ctx == curr_run_ctx);
```
```suggestion
    serverAssert(run_ctx == curr_run_ctx);
```

sundb commented on src/script.c:
```diff
@@ -0,0 +1,465 @@
+}
+
+static void enterScriptTimedoutMode(scriptRunCtx *run_ctx) {
+	serverAssert(run_ctx == curr_run_ctx);
```
```suggestion
    serverAssert(run_ctx == curr_run_ctx);
```

sundb commented on src/functions.h:
```diff
@@ -0,0 +1,126 @@
+void functionsCtxFree(functionsCtx *functions_ctx);
+void functionsCtxClear(functionsCtx *functions_ctx);
+void functionsCtxSwapWithCurrent(functionsCtx *functions_ctx);
+
```
One extra blank line.

oranagra commented:
For the record, we discussed the persistence issue in a core-team meeting today, we concluded that unlike ACL (which is more an admin related feature), Functions are an application related feature, and should be persisted together with the data.
We do need to resolve the problem with redis cluster in some way (possibly propagating them on the cluster bus, or adding some tools that will allow cluster admins to easily load them on all nodes), we will look into that in a followup PR.

oranagra reviewed (approved):
recent changes LGTM.
(other than pending discussion about FUNCTION INFO vs FUNCTION LIST)

madolson commented:
There was one other point in the meeting last night that was mentioned that I forgot to circle back to, which was the ephemeral use cases. Like, spinning up a cache but not using replication, we should make it easy for functions to be available there. It was mentioned that this is not common, but I disagree and think that use case is extremely common. 

I also think clusterbus propagation of functions isn't a Redis 7 feature, so we should aim for something that solves cluster + ephemeral workloads, which is probably just passing in a startup config and maybe some rewrite functionality (or listing all the functions with their descriptions) so they can be loaded on a new cluster node.

sundb commented on src/functions.c:
```diff
@@ -259,28 +253,47 @@ void functionsInfoCommand(client *c) {
+
+    functionInfo *fi = dictFetchValue(functions_ctx->functions, function_name->ptr);
+    if (!fi) {
+        addReplyError(c, "Function already exists");
```
Should it be `Function doesn't exist`?

MeirShpilraien commented:
Squashed all reviews commits to their relevant location, kept only the original 5 commits.

oranagra commented:
triggered full CI https://github.com/redis/redis/actions/runs/1529820590

madolson reviewed (approved):
Talked with @oran over slack, and it seems like we're definitely going to make sure there is another strategy for making functions loadable at startup/cluster mode. So we can follow up on that issue.

MeirShpilraien commented:
Loading Functions on cluster and/or startup time is discussed here: https://github.com/redis/redis/issues/9899

MeirShpilraien commented:
Code sharing on functions is discussed here: https://github.com/redis/redis/issues/9906
